rfd/XXXX-cloud-agent-upgrades.md

```diff
@@ -0,0 +1,318 @@
+RandomizedDelaySec=90
+
+[Install]
+WantedBy=teleport-agent.service
+```
+
+The exact time values are TBD. We want upgrades to be reliably and frequently discovered, but don't want to end up overloading
+repositories. This is why we aren't just using `OnActiveSec=1`.  If teleport is in *too* tight of a crash loop, we don't want
+to just end up hitting the repos in a tight loop. This will definitely take some fine-tuning, and we may want to customize the
+`StartLimitIntervalSec` and `StartLimitBurst` of the main teleport agent unit.
+
+---
+
+TODO: apt/yum channel and artifact promotion details
+
+
+### Automatic/Guided Install Changes
+
+Teleport provides two kinds of automatic/guided install mechanism for adding the teleport agent software to existing
+machines. There are "discovery installers" which are scripts used by ec2 auto-discovery (and possibly by other forms of
+server auto-discovery in the future) to install teleport agents automatically, and there is the `node-join` script template
+used by the web UI to serve custom installers during the interactive agent-adding process.
+
+To ensure a seamless experience, we will need both of these install scripts to be capable of being parameterized or overridden s.t.
+they target `cloud-stable` when the cluster is configured for continuous agent upgrades.
+
+The current implementation of the `node-join` script template relies upon direct downloads from `get.gravitational.com`. It would
+be preferable to force use of a package manager when installing `cloud-stable`, as that would let us improve standardization, minimize
+the amount of custom logic in the download path, and (most importantly) reduce the complexity of the periodic upgrade operation,
+which is the most sensitive part of this proposal.
+
+I believe the best possible outcome would be to unify the core logic of the "discovery", "join", and "upgrade" scripts into a single
+cross-platform function so that we have a single source of truth for package manager discovery and invocation going forward. The
+feasability of this will require further investigation.
+
+
+### Summary of Changes
+
+- Agents will expose a flags to enable the VM and k8s variants of the agent upgrade behavior.
+
+- Package repositories will expose a `cloud-stable` variant of packages which will lag slightly behind "latest"
+and be managed indipendently of the standard release process.
+
+- A new `cloud-stable` helm chart (or option) will be dstributed.
+
+- A new `teleport-upgrade` script and accompanying systemd timer unit will be distributed as part of the
+`cloud-stable` apt/yum packages.
+
+- Teleport auth servers will be modified to be able to serve the maintenance window to agents.
+
+- Cloud clusters will have the option to track `cloud-stable` rather than a specific major version.
+
+- Teleport discovery and join features will be modified to serve appropriate install scripts when `cloud-stable`
+tracking is enabled for a cluster.
+
```

Comment:
How will rollbacks be handled?

Reply:
The intent is to have rollbacks treated as equivalent to upgrades (just have a lower version become the new target of cloud-stable instead of a higher version).